test/Generators.hs

```diff
@@ -15,17 +15,48 @@ import qualified Hedgehog.Range as Range
 
 -- DISCREPANCY: 'jq' allows both 000 in input and its own number literals,
 -- but JSON spec does not. We'll stay closest to 'jq' by doing this, too.
+--
 
--- First, an unsized Gen of JSON numbers.
+-- | Helper Generators
+signGen :: Gen Text
+signGen = Gen.element ["+", "-", ""]
 
+signGen' :: Gen Text
+signGen' = Gen.element ["-", ""]
+
+integerGen :: Gen Text
+integerGen = Text.pack . show <$>
+      (Gen.integral $ Range.exponential 1 (10^309))
+
+expGen :: Gen Text
+expGen = Text.pack . show <$>
+  (Gen.integral (Range.linearFrom 0 (-10^16) (10^16)))
+
+decBefore :: Gen Text
+decBefore = Gen.constant "." <> integerGen
+
+decAfter :: Gen Text
+decAfter = integerGen <> Gen.constant "."
+
+-- | Number type generators
 numberGen :: Gen Text
-numberGen = sign <> integer <> commaPart
+numberGen = signGen' <> integerGen
+
+scientificNumberGen :: Gen Text
+scientificNumberGen = base <> e <> signGen <> expGen
   where
-    sign = Gen.element ["", "-"]
-    integer = Text.singleton <$> Gen.element ['0'..'9']
-    commaPart = Gen.element [ "..." ]
+    base = Text.singleton <$> Gen.element ['0'..'9']
+    e = Gen.element ["e", "E"]
```

Comment:
It seems that `scientificNumberGen` does not currently generate the combination of an arbitrarily large integer part and an arbitrarily large fraction. I'd like it to also hit these edge cases because imprecision happens to `double`s when any one part grows. So when this generator is eventually used to show the discrepancies between how jq and hs-jq parse numbers, I'd like as many red flags as possible.

Reply:
Good catch